gapis/api/vulkan/frame_loop.go

```diff
@@ -126,129 +139,237 @@ func newFrameLoop(ctx context.Context, c *capture.GraphicsCapture, numInitialCmd
 		imageChanged:   make(map[VkImage]bool),
 		imageDestroyed: make(map[VkImage]bool),
 		imageToBackup:  make(map[VkImage]VkImage),
-	}
 
-	f.loopStartCmd, f.loopEndCmd = f.getLoopStartAndEndCmd(ctx, Cmds)
-
-	return f
+		loopTerminated:      false,
+		lastObservedCommand: api.CmdNoID,
+	}
 }
 
-func (f *frameLoop) Transform(ctx context.Context, id api.CmdID, cmd api.Cmd, out transform.Writer) {
+func (f *frameLoop) Transform(ctx context.Context, cmdId api.CmdID, cmd api.Cmd, out transform.Writer) {
+
 	ctx = log.Enter(ctx, "FrameLoop Transform")
+	log.D(ctx, "FrameLoop: looping from %v to %v.", f.loopStartIdx, f.loopEndIdx)
+
+	// Lets capture and update the last observed frame from f. From this point on use the local lastObservedCommand variable.
+	lastObservedCommand := f.lastObservedCommand
+	f.lastObservedCommand = cmdId
 
-	if cmd == f.loopStartCmd {
-		log.D(ctx, "Loop: start loop at frame %v, id %v, cmd %v.", f.frameNum, id, cmd)
-		f.detectChangedResource(ctx, out.State())
+	if lastObservedCommand != api.CmdNoID && lastObservedCommand > api.CmdID.Real(cmdId) {
+		log.F(ctx, true, "FrameLoop: expected next observed command ID to be >= last observed command ID")
+	}
+
+	// Walk the frame count forwards if we just hit the end of one.
+	if _, ok := cmd.(*VkQueuePresentKHR); ok {
+		f.frameNum++
+	}
 
-		st := GetState(out.State())
-		sb := st.newStateBuilder(ctx, newTransformerOutput(out))
-		defer sb.ta.Dispose()
+	// Are we before the loop or just at the start of it?
+	if lastObservedCommand == api.CmdNoID {
+
+		// This is the start of the loop.
+		if api.CmdID.Real(cmdId) == f.loopStartIdx {
```

Comment:
I expect this to be `api.CmdID.Real(cmdId) >= f.loopStartIdx || loopStarted` or something similar.

Reply:
I wanted to write the loops so no logical terms were contained in more than one "(else) if" in the branch structure. I also wanted to avoid representing loopStarted as a bool as it opens up the possibility of having loopEnded = true, loopStarted = false, which is a logically inconsistent state. 

Do you find the control flow hard to read? I guess ideally the blocks would be in order of before/start/inside/end/after, but that wasn't possible without contravening the above.

Reply:
My main concern was what happens if f.loopStartIdx does not end up being any of the actual commands. Unless I am missing something, this seems possible?

Reply:
Whoops! Fixed with >=

Really good catch!